File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [[float('inf')]*(amount+1) for _ in range(len(coins)+1)]
        for l in range(len(dp)):
            for amt in range(amount+1):
                if amt==0: dp[l][amt] = 0
                elif l!=0:
                    if amt>=coins[l-1]: dp[l][amt] = min(dp[l-1][amt], 1 + dp[l][amt-coins[l-1]])
                    else: dp[l][amt] = dp[l-1][amt]
        return -1 if dp[-1][-1] == float('inf') else dp[-1][-1]

File: 0322-coin-change/0322-coin-change.java
class Solution {
    int[][] memo;
    public int coinChange(int[] coins, int amount) {
        memo = new int[coins.length+1][amount+1];
        for(int[] m1: memo) Arrays.fill(m1, -1);
        
        help(coins.length, coins, amount);
        return memo[coins.length][amount]>10000 ? -1 : memo[coins.length][amount];
    }
    private int help(int l, int[] coins, int amt){
        if(l==0) return amt==0? 0: 100001;
        if(memo[l][amt] != -1) return memo[l][amt];
        memo[l][amt] = help(l-1, coins, amt) ;
        if(amt>=coins[l-1]) memo[l][amt] = Math.min(memo[l][amt], 1 + help(l, coins, amt-coins[l-1]));
        return memo[l][amt];
    }
}